homeassistant/components/octoprint/__init__.py

```diff
@@ -202,3 +232,18 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
         hass.data[DOMAIN].pop(entry.entry_id)
 
     return unload_ok
+
+
+def async_get_client_for_service_call(
+    hass: HomeAssistant, call: ServiceCall
+) -> OctoprintClient:
+    """Get the client related to a service call (by device ID)."""
+    device_id = call.data[CONF_DEVICE_ID]
+    device_registry = dr.async_get(hass)
+
+    if device_entry := device_registry.async_get(device_id):
+        for entry_id in device_entry.config_entries:
+            if data := hass.data[DOMAIN].get(entry_id):
+                return cast(OctoprintClient, data["client"])
+
+    raise ValueError(f"No client for device ID: {device_id}")
```

Comment:
This should be a `ServiceValidationError`, ideally with a tranlsation in strings.json explaining what went wrong.

Like this:
https://github.com/home-assistant/core/blob/ef5d46c79c00f92425c51b84d8f1b093be27f3c4/homeassistant/components/vicare/climate.py#L305-L311